packages/app-router/src/hooks.tsx

```diff
@@ -0,0 +1,620 @@
+  ): Promise<U> {
+    // Generate a unique name for the modal component.
+    const name = `Modal_${idCounter}`;
+    idCounter++;
+
+    // Register the modal component with React Native Navigation.
+    Navigation.registerComponent(
+      name,
+      () => props => {
+        const {componentId, resolve, reject, ...passProps} = props;
+
+        return (
+          <ComponentIdContext.Provider value={componentId}>
+            <ModalContext.Provider value={{resolve, reject, data: passProps}}>
+              <Component />
+            </ModalContext.Provider>
+          </ComponentIdContext.Provider>
+        );
+      },
+      () => Component,
+    );
+
+    // Return a promise that resolves or rejects based on modal dismissal.
+    return new Promise((res, rej) => {
+      // Define the resolve function that dismisses the modal and resolves the promise.
+      function resolve(componentId: string) {
+        return (data: U) => {
+          res(data);
+
+          Navigation.dismissModal(componentId);
+        };
+      }
+
+      // Define the reject function that dismisses the modal and rejects the promise.
+      function reject(componentId: string) {
+        return () => {
+          rej();
+
+          Navigation.dismissModal(componentId);
+        };
+      }
+
+      // Show the modal using React Native Navigation with the provided options.
+      Navigation.showModal({
+        stack: {
+          children: [
+            {
+              component: {
+                name,
+                passProps: {
+                  ...data,
+                  resolve,
+                  reject,
+                },
+                options: {
+                  ...((Component as any).options ?? {}),
+                  ...options,
+                },
+              },
+            },
+          ],
+        },
+      });
+    });
+  }
+
+  return {
+    open,
+    push,
+    pop,
+    popToRoot,
+    popTo,
+    setStackRoot,
+    showModal,
+  };
+}
+
+/**
+ * A custom hook that manages deep linking by listening for URL changes
+ * and navigating accordingly using a navigator.
+ *
+ * @example
+ * function App() {
+ *   useLinking();
+ *
+ *   return <YourAppComponent />;
+ * }
+ *
+ * // This hook will automatically handle incoming deep links and navigate
+ * // based on the URL, such as `yourapp://path?query=param`.
+ */
+export function useLinking() {
+  const navigator = useNavigator();
+
+  /**
+   * Handles the URL passed from deep linking and navigates to the correct screen.
+   *
+   * @param {Object} params - Parameters object.
+   * @param {string | null} params.url - The URL to be processed.
+   *
+   * @example
+   * callback({ url: 'yourapp://profile?user=123' });
+   */
+  function callback({url}: {url: string | null}) {
+    if (!url) return;
+
+    try {
+      const {pathname, query} = urlParse(url, true);
+
+      // Navigates to the parsed URL's pathname and search params
+      navigator.open(pathname + query);
```

Comment:
I believe query is a Record<string, string> here, do we want to pass the query as a prop here?

Reply:
correct we actually need to pass `false` i.e. leave it empty to urlParse

Reply:
misread typings, query is string